deepmd/pt/utils/stat.py

```diff
@@ -36,55 +34,168 @@
+    - A list of dicts, each of which contains data from a system.
     """
     lst = []
     log.info(f"Packing data for statistics from {len(datasets)} systems")
-    for i in range(len(datasets)):
-        sys_stat = {}
-        with torch.device("cpu"):
-            iterator = iter(dataloaders[i])
-            numb_batches = min(nbatches, len(dataloaders[i]))
-            for _ in range(numb_batches):
-                try:
-                    stat_data = next(iterator)
-                except StopIteration:
-                    iterator = iter(dataloaders[i])
-                    stat_data = next(iterator)
-                for dd in stat_data:
-                    if stat_data[dd] is None:
-                        sys_stat[dd] = None
-                    elif isinstance(stat_data[dd], torch.Tensor):
-                        if dd not in sys_stat:
-                            sys_stat[dd] = []
-                        sys_stat[dd].append(stat_data[dd])
-                    elif isinstance(stat_data[dd], np.float32):
-                        sys_stat[dd] = stat_data[dd]
-                    else:
-                        pass
+    total_element_types = set()
+    global_element_counts = {}
+    collect_ele = defaultdict(int)
+    if datasets[0].mixed_type:
+        if enable_element_completion:
+            log.info(
+                f'Element check enabled. '
+                f'Verifying if frames with elements meet the set of {min_frames_per_element_forstat}.'
+            )
+        else:
+            log.info("Element completion is disabled. Skipping missing element handling.")
+
+    def process_batches(dataloader, sys_stat):
+        """Process batches from a dataloader to collect statistics."""
+        iterator = iter(dataloader)
+        numb_batches = min(nbatches, len(dataloader))
+        for _ in range(numb_batches):
+            try:
+                stat_data = next(iterator)
+            except StopIteration:
+                iterator = iter(dataloader)
+                stat_data = next(iterator)
+            for dd in stat_data:
+                if stat_data[dd] is None:
+                    sys_stat[dd] = None
+                elif isinstance(stat_data[dd], torch.Tensor):
+                    if dd not in sys_stat:
+                        sys_stat[dd] = []
+                    sys_stat[dd].append(stat_data[dd])
+                elif isinstance(stat_data[dd], np.float32):
+                    sys_stat[dd] = stat_data[dd]
+                else:
+                    pass
+
+    def process_with_new_frame(sys_indices, newele_counter):
+        for sys_info in sys_indices:
+            sys_index = sys_info['sys_index']
+            frames = sys_info['frames']
+            sys = datasets[sys_index]
+            for frame in frames:
+                newele_counter += 1
+                if not newele_counter > min_frames_per_element_forstat:
+                    frame_data = sys.__getitem__(frame)
```

Comment:
Better to add a check to ensure that this frame really contains this missing element. Use `assert`.